src/main/java/com/direwolf20/buildinggadgets/common/tools/ToolRenders.java

```diff
@@ -64,304 +69,176 @@ public static void updateInventoryCache() {
-            GlStateManager.blendFunc(GL11.GL_CONSTANT_ALPHA, GL11.GL_ONE_MINUS_CONSTANT_ALPHA);
-            GlStateManager.translate(-doubleX, -doubleY, -doubleZ);//The render starts at the player, so we subtract the player coords and move the render to 0,0,0
-            GlStateManager.translate(pos.getX(), pos.getY(), pos.getZ());//Now move the render position to the coordinates we want to render at
-            GlStateManager.rotate(-90.0F, 0.0F, 1.0F, 0.0F); //Rotate it because i'm not sure why but we need to
-            GlStateManager.translate(-0.005f, -0.005f, 0.005f);
-            GlStateManager.scale(1.01f, 1.01f, 1.01f);
+        IBlockState renderBlockState = getToolBlock(heldItem);
+
+        //Don't render anything if there is no block selected (Air)
+        if (renderBlockState == Blocks.AIR.getDefaultState())
+            return;
+
+        //Build a list of coordinates based on the tool mode and range
+        if (coordinates.size() == 0 && lookingAt != null)
+            coordinates = BuildingModes.collectPlacementPos(player.world, player, lookingAt.getBlockPos(), lookingAt.sideHit, heldItem, lookingAt.getBlockPos());
+
+        // Figure out how many of the block we're rendering we have in the inventory of the player.
+        ItemStack itemStack = ToolRenders.Utils.getSilkDropIfPresent(player.world, renderBlockState, player);
+
+        // Check if we have the blocks required
+        long hasBlocks = InventoryManipulation.countItem(itemStack, player, cacheInventory);
+        hasBlocks += InventoryManipulation.countPaste(player);
+
+        int hasEnergy = ToolRenders.Utils.getStackEnergy(heldItem, player);
+
+        // Prepare the fake world -- using a fake world lets us render things properly, like fences connecting.
+        Set<BlockPos> coords =  new HashSet<>(coordinates);
+        fakeWorld.setWorldAndState(player.world, renderBlockState, coords);
+
+        GlStateManager.pushMatrix();
+        ToolRenders.Utils.stateManagerPrepareBlend();
+
+        List<BlockPos> sortedCoordinates = Sorter.Blocks.byDistance(coordinates, player); //Sort the coords by distance to player.
+
+        for (BlockPos coordinate : sortedCoordinates) {
+            ToolRenders.Utils.stateManagerPrepare(playerPos, coordinate, false, null);
+            GL14.glBlendColor(1F, 1F, 1F, 0.55f); //Set the alpha of the blocks we are rendering
+
+            if (fakeWorld.getWorldType() != WorldType.DEBUG_ALL_BLOCK_STATES)
+                state = renderBlockState.getActualState(fakeWorld, coordinate);
+
+            mc.getBlockRendererDispatcher().renderBlockBrightness(state, 1f);//Render the defined block
+
+            GlStateManager.popMatrix();
+        }
+
+        for (BlockPos coordinate : coordinates) { //Now run through the UNSORTED list of coords, to show which blocks won't place if you don't have enough of them.
+            ToolRenders.Utils.stateManagerPrepare(playerPos, coordinate, true, 0.005f);
```

Comment:
I presume that this does a pushMatrix? I find it rather confusing not to see push and pop...

Could you maybe move the push out even if it is duplicate code then?

Reply:
Sure. I'll put the push back :)